source/common/config/remote_data_fetcher.cc

```diff
@@ -0,0 +1,82 @@
+
+#include "common/common/enum_to_int.h"
+#include "common/common/hex.h"
+#include "common/crypto/utility.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+namespace Envoy {
+namespace Config {
+namespace DataFetcher {
+
+RemoteDataFetcher::RemoteDataFetcher(Upstream::ClusterManager& cm,
+                                     const ::envoy::api::v2::core::HttpUri& uri,
+                                     const std::string& content_hash,
+                                     RemoteDataFetcherCallback& callback)
+    : cm_(cm), uri_(uri), content_hash_(content_hash), callback_(callback) {}
+
+RemoteDataFetcher::~RemoteDataFetcher() { cancel(); }
+
+void RemoteDataFetcher::cancel() {
+  if (request_ && !complete_) {
+    request_->cancel();
+    ENVOY_LOG(debug, "fetch remote data [uri = {}]: canceled", uri_.uri());
+  }
+
+  request_ = nullptr;
+}
+
+void RemoteDataFetcher::fetch() {
+  complete_ = false;
+
+  Http::MessagePtr message = Http::Utility::prepareHeaders(uri_);
+  message->headers().insertMethod().value().setReference(Http::Headers::get().MethodValues.Get);
+  ENVOY_LOG(debug, "fetch remote data from [uri = {}]: start", uri_.uri());
+  request_ = cm_.httpAsyncClientForCluster(uri_.cluster())
+                 .send(std::move(message), *this,
+                       Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(
+                           DurationUtil::durationToMilliseconds(uri_.timeout()))));
+}
+
+void RemoteDataFetcher::onSuccess(Http::MessagePtr&& response) {
+  complete_ = true;
+  const uint64_t status_code = Http::Utility::getResponseStatus(response->headers());
+  if (status_code == enumToInt(Http::Code::OK)) {
+    ENVOY_LOG(debug, "fetch remote data [uri = {}]: success", uri_.uri());
+    if (response->body()) {
+      const auto len = response->body()->length();
+      const auto content_hash =
+          Hex::encode(Envoy::Common::Crypto::Utility::getSha256Digest(*response->body()));
+
+      if (content_hash_ != content_hash) {
+        ENVOY_LOG(debug, "fetch remote data [uri = {}]: data is invalid", uri_.uri());
+        callback_.onFailure(Failure::InvalidData);
+      } else {
+        const auto body = std::string(static_cast<char*>(response->body()->linearize(len)), len);
+        callback_.onSuccess(body);
+      }
+    } else {
+      ENVOY_LOG(debug, "fetch remote data [uri = {}]: body is empty", uri_.uri());
+      callback_.onFailure(Failure::Network);
+    }
+  } else {
+    ENVOY_LOG(debug, "fetch remote data [uri = {}]: response status code {}", uri_.uri(),
+              status_code);
+    callback_.onFailure(Failure::Network);
+  }
+
+  request_ = nullptr;
+}
+
+void RemoteDataFetcher::onFailure(Http::AsyncClient::FailureReason reason) {
+  ENVOY_LOG(debug, "fetch remote data [uri = {}]: network error {}", uri_.uri(), enumToInt(reason));
+  complete_ = true;
+  callback_.onFailure(Failure::Network);
+
+  request_ = nullptr;
```

Comment:
won't be executed on exception

Reply:
Good catch. I will move it before callback.